src/datasource.ts

```diff
@@ -16,23 +16,107 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { DataSourceInstanceSettings, ScopedVars } from '@grafana/data';
-import { DataSourceWithBackend, getTemplateSrv } from '@grafana/runtime';
+import { DataFrame, DataSourceInstanceSettings, ScopedVars, TimeRange, toDataFrame } from '@grafana/data';
+import { BackendSrv, DataSourceWithBackend, FetchResponse, getBackendSrv, getTemplateSrv } from '@grafana/runtime';
 import { PixieDataSourceOptions, PixieDataQuery } from './types';
 
+interface ClusterMeta {
+  id: string;
+  name: string;
+}
+
 export class DataSource extends DataSourceWithBackend<PixieDataQuery, PixieDataSourceOptions> {
+  backendServ?: BackendSrv;
+
   constructor(instanceSettings: DataSourceInstanceSettings<PixieDataSourceOptions>) {
     super(instanceSettings);
+    this.backendServ = getBackendSrv();
   }
 
   applyTemplateVariables(query: PixieDataQuery, scopedVars: ScopedVars) {
+    const pxlScript = query?.queryBody?.pxlScript;
     return {
       ...query,
-      pxlScript: query.pxlScript
-        ? getTemplateSrv().replace(query.pxlScript, {
-            ...scopedVars,
-          })
-        : '',
+      queryBody: {
+        pxlScript: pxlScript
+          ? getTemplateSrv().replace(pxlScript, {
+              ...scopedVars,
+            })
+          : '',
+      },
+    };
+  }
+
+  async fetchMetricNames(options: any): Promise<FetchResponse | void> {
+    let refId = 'tempvar';
+    if (options && options.variable && options.variable.name) {
+      refId = options.variable.name;
+    }
+
+    const range = options?.range as TimeRange;
+    const interpolatedQuery = {
+      refId: refId,
```

Comment:
```suggestion
      refId,
```
Object shorthand syntax is nice for stuff like this